server/aesdsocket.c

```diff
@@ -274,104 +553,45 @@ int main(int argc, char *argv[])
             syslog(LOG_DEBUG,"Accepted connection from %s", ip_addr);
         }
 
-        char *data_buf = (char*)calloc(initial_alloc_size,sizeof(char));
-        if(NULL == data_buf)
+        //set the members of single linked list
+        datap = (slist_data_t*)malloc(sizeof(slist_data_t));
+        SLIST_INSERT_HEAD(&head,datap,entries);
+        datap->thread_values.is_thread_finished = false;
+        datap->thread_values.my_mutex = &mutex;
+        datap->thread_values.fd = data_file_fd;
+        datap->thread_values.client_sock = client_sockfd;
+
+        //create the thread
+        if(pthread_create(&(datap->thread_values.my_thread),NULL,&thread_routine,(void*)&(datap->thread_values)) != 0)
         {
-            perror("calloc()");
+            perror("pthread_create()");
             error_handler();
             exit(ERROR_CODE);
         }
 
         /*
-         * For completing any open connection operations i.e., resv() and send(), need to block the signal using sigprocmask()
-         * Blocking doesn't mean we miss the signal, the signal waits until its unblocked and gets serviced by kernel. 
-         */
-		if( -1 == sigprocmask(SIG_BLOCK, &sock_set, NULL))
-        {
-			perror("sigprocmask()");
-			error_handler();
-			exit(ERROR_CODE);
-        }
-
-        while((bytes_read = recv(client_sockfd,data_buf+present_location,initial_alloc_size,0)) > 0)
-        {
-
-            if(-1 == bytes_read)
+        * it iterates through a singly-linked list using SLIST_FOREACH_SAFE, and for each element in the list, 
+        * it waits for the associated thread (referenced by my_thread) to finish using pthread_join. 
+        * If the thread has completed (as indicated by is_thread_finished), it removes the element from the list and frees the associated memory.
+        */
+        SLIST_FOREACH_SAFE(datap,&head,entries,loop)
+		{
+			if(pthread_join(datap->thread_values.my_thread,NULL) !=0)
```

Comment:
@Rishikesh38 by putting `pthread_join` first I think you are always going to only run 1 thread at a time.  The intent was to only join completed threads here (check for `is_thread_finished` first) and therefore allow running more than 1 thread at a given time.